x/mongo/driver/topology/connection_test.go

```diff
@@ -833,6 +833,134 @@ func TestConnection(t *testing.T) {
 				t.Errorf("LocalAddresses do not match. got %v; want %v", got, want)
 			}
 		})
+
+		t.Run("pinning", func(t *testing.T) {
+			makeMultipleConnections := func(t *testing.T, numConns int) (*pool, []*Connection) {
+				t.Helper()
+
+				addr := address.Address("")
+				pool, err := newPool(poolConfig{Address: addr})
+				assert.Nil(t, err, "newPool error: %v", err)
+
+				err = pool.sem.Acquire(context.Background(), int64(numConns))
+				assert.Nil(t, err, "error acquiring semaphore: %v", err)
+
+				conns := make([]*Connection, 0, numConns)
+				for i := 0; i < numConns; i++ {
+					conn, err := newConnection(addr)
+					assert.Nil(t, err, "newConnection error: %v", err)
+					conn.pool = pool
+					conns = append(conns, &Connection{connection: conn})
+				}
+				return pool, conns
+			}
+			makeOneConnection := func(t *testing.T) (*pool, *Connection) {
+				t.Helper()
+
+				pool, conns := makeMultipleConnections(t, 1)
+				return pool, conns[0]
+			}
+
+			assertPoolPinnedStats := func(t *testing.T, p *pool, cursorConns, txnConns uint64) {
+				t.Helper()
+
+				assert.Equal(t, cursorConns, p.pinnedCursorConnections, "expected %d connections to be pinned to cursors, got %d",
+					cursorConns, p.pinnedCursorConnections)
+				assert.Equal(t, txnConns, p.pinnedTransactionConnections, "expected %d connections to be pinned to transactions, got %d",
+					txnConns, p.pinnedTransactionConnections)
+			}
+
+			t.Run("cursors", func(t *testing.T) {
+				pool, conn := makeOneConnection(t)
+				err := conn.PinToCursor()
+				assert.Nil(t, err, "PinToCursor error: %v", err)
+				assertPoolPinnedStats(t, pool, 1, 0)
+
+				err = conn.UnpinFromCursor()
+				assert.Nil(t, err, "UnpinFromCursor error: %v", err)
+
+				err = conn.Close()
+				assert.Nil(t, err, "Close error: %v", err)
+				assertPoolPinnedStats(t, pool, 0, 0)
+			})
+			t.Run("transactions", func(t *testing.T) {
+				pool, conn := makeOneConnection(t)
+				err := conn.PinToTransaction()
+				assert.Nil(t, err, "PinToTransaction error: %v", err)
+				assertPoolPinnedStats(t, pool, 0, 1)
+
+				err = conn.UnpinFromTransaction()
+				assert.Nil(t, err, "UnpinFromTransaction error: %v", err)
+
+				err = conn.Close()
+				assert.Nil(t, err, "Close error: %v", err)
+				assertPoolPinnedStats(t, pool, 0, 0)
+			})
+			t.Run("pool is only updated for first reference", func(t *testing.T) {
+				pool, conn := makeOneConnection(t)
+				err := conn.PinToTransaction()
+				assert.Nil(t, err, "PinToTransaction error: %v", err)
+				assertPoolPinnedStats(t, pool, 0, 1)
+
+				err = conn.PinToCursor()
```

Comment:
Oh, maybe this answers my previous question. You may need to pin the same connection to a transaction and a cursor?

Reply:
Yep.

Reply:
Ah right, that's the double counting protection.